dataPipelines/gc_thumbnails/utils.py

```diff
@@ -1,26 +1,48 @@
 from pathlib import Path
 import fitz
 import typing as t
+import multiprocessing
+from concurrent.futures import ProcessPoolExecutor
+
 
 class ThumbnailsCreator:
     def __init__(
         self,
         input_directory: t.Union[str, Path],
-        output_directory: t.Union[str, Path]
+        output_directory: t.Union[str, Path],
+        max_threads: int
     ):
         self.input_directory = Path(input_directory).absolute()
         self.output_directory = Path(output_directory).absolute()
+        self.max_threads = max_threads
 
-    def generate_thumbnails(self):
+    def process_directory(self):
         print('\nGenerating Thumbnails\n')
         self.output_directory.mkdir(exist_ok=True)
-        for file_path in self.input_directory.glob('*.pdf'):
-            doc = fitz.open(str(file_path))
-            page = doc.loadPage(0)  # number of page
-            pix = page.getPixmap()
-            output = Path(self.output_directory, file_path.with_suffix('.png').name)
-            pix.writePNG(str(output))
-
-            print("wrote PNG: ", output)
+        paths = self.input_directory.glob('*.pdf')
+
+        # if we use all available resources
+        # NOT recommended. This uses all computing power at once, will probably crash if big directory
+        if self.max_threads < 0:
```

Comment:
all the logic for validating `max_threads` param should be in `__init__` method, since that's when the parameter is passed.

Reply:
great call. changed